build_tools/python/e2e_test_framework/definitions/iree_definitions.py

```diff
@@ -161,49 +161,197 @@ def materialize_import_flags(self,
 }
 
 
-@serialization.serializable
+@serialization.serializable(type_key="iree_imported_models",
+                            id_field="composite_id")
 @dataclass(frozen=True)
 class ImportedModel(object):
   """Describes an imported MLIR model."""
+  composite_id: str
   model: common_definitions.Model
   import_config: ImportConfig
 
-  def composite_id(self):
-    return unique_ids.hash_composite_id([self.model.id, self.import_config.id])
-
   @staticmethod
   def from_model(model: common_definitions.Model):
     config = MODEL_SOURCE_TO_DEFAULT_IMPORT_CONFIG_MAP.get(model.source_type)
     if config is None:
       raise ValueError(f"Unsupported model source type: {model.source_type}.")
 
-    return ImportedModel(model=model, import_config=config)
+    composite_id = unique_ids.hash_composite_id([model.id, config.id])
+    return ImportedModel(composite_id=composite_id,
+                         model=model,
+                         import_config=config)
 
 
-@serialization.serializable
+@serialization.serializable(type_key="iree_module_generation_configs",
+                            id_field="composite_id")
 @dataclass(frozen=True)
 class ModuleGenerationConfig(object):
   """Describes a compile target to generate the module."""
+  composite_id: str
   imported_model: ImportedModel
   compile_config: CompileConfig
+  # Full list of flags to compile with, derived from sub-components, with
+  # unmaterialized placeholders. Allows the compile flags to be persisted and
+  # decouple from the generation code. Also serves as useful information in the
+  # serialized JSON.
+  composite_flags: List[str]
```

Comment:
I'm confused by the "composite" aspect of this. The flags can only be for one tool and AFAICT they're all compiler flags. If they were for two separate tools, I'd want them separate anyway

Reply:
It's the full flag list for `iree-compile` (including args for driver, target architecture, ..). Maybe `full_flags` is a better name?

Reply:
Well why not "compile_flags"? But also, I think it would make more sense for these to be part of the compile_config, no?

Reply:
It is because part of the compile flags come from the model (actually only one `--iree-input-type=` ...). `compile_flags` SGTM

Reply:
Ah, ok that was the thing I was missing. It's a flag for the compiler but it's determined by the model. Makes sense, thanks